src/cryptonote_basic/cryptonote_format_utils.cpp

```diff
@@ -1858,21 +1858,24 @@ namespace cryptonote
   //---------------------------------------------------------------
   crypto::hash get_block_content_hash(const block& b)
   {
+    // The miner transaction goes first in the Merkle tree so that mining software can update nonces
+    // within, and then the block content hash can be recalculated by re-hashing the left-hand side
+    // of the tree. The number of FCMP tree layers and FCMP tree root go in the following slots so
+    // that they are at a fixed index.
+
     std::vector<crypto::hash> hashes;
     hashes.reserve(1 + 1 + 1 + b.tx_hashes.size());
-    // 1. n tree layers in FCMP++ tree
-    static_assert(sizeof(crypto::hash) >= sizeof(uint8_t), "crypto::hash is too small");
+    // 1. Miner tx
+    hashes.push_back(get_transaction_hash(b.miner_tx));
     if (b.major_version >= HF_VERSION_FCMP_PLUS_PLUS)
+    {
+      // 2. n tree layers in FCMP++ tree
+      static_assert(sizeof(crypto::hash) >= sizeof(uint8_t), "crypto::hash is too small");
       hashes.push_back(crypto::hash{static_cast<char>(b.fcmp_pp_n_tree_layers)});
-    // 2. FCMP++ tree root
-    static_assert(sizeof(crypto::hash) == sizeof(crypto::ec_point), "expected sizeof hash == sizeof ec_point");
-    if (b.major_version >= HF_VERSION_FCMP_PLUS_PLUS)
-      hashes.push_back((crypto::hash&)b.fcmp_pp_tree_root);
-    // 3. Miner tx
-    crypto::hash h = null_hash;
-    size_t bl_sz = 0;
-    CHECK_AND_ASSERT_THROW_MES(get_transaction_hash(b.miner_tx, h, bl_sz), "Failed to calculate transaction hash");
-    hashes.push_back(h);
+      // 3. FCMP++ tree root
+      static_assert(sizeof(crypto::hash) == sizeof(crypto::ec_point), "expected sizeof hash == sizeof ec_point");
+      memcpy(&hashes.emplace_back(), &b.fcmp_pp_tree_root, sizeof(crypto::hash));
```

Comment:
This code looks very weird to my C++ brain... Too much stuff is squeezed into one line. I would do something like this
```
crypto::hash& h = hashes.emplace_back();
static_assert(sizeof(h.data) == sizeof(b.fcmp_pp_tree_root.data), "data size mismatch");
memcpy(h.data, b.fcmp_pp_tree_root.data, sizeof(h.data));
```